Add db recovery methods

sakridge opened the pull request:
#### Problem

Blockstore encounters errors opening the db because wal/sst corruption.

#### Summary of Changes

Allow for skipping corrupted wal entries.

Fixes #10015

mvines commented on validator/src/main.rs:
```diff
@@ -843,6 +844,17 @@ pub fn main() {
+                .takes_value(true)
+                .help(
+                    "Mode to recovery the ledger db write ahead log. \
+                    Options: tolerate_corrupted_tail_records, absolute_consistency,\
```
Try using `.possible_values()` to allow clap to document and validate the user's input for ya:
https://github.com/solana-labs/solana/blob/4b93a7c1f6859d3771d5e6a5ec1a03f763eb12b1/dos/src/main.rs#L110-L118

sakridge replied:
oh cool

t-nelson commented on validator/src/main.rs:
```diff
@@ -869,6 +884,18 @@ pub fn main() {
+        matches
+            .value_of("wal_recovery_mode")
+            .map(|recovery_mode| match recovery_mode {
+                "tolerate_corrupted_tail_records" => {
+                    BlockstoreRecoveryMode::TolerateCorruptedTailRecords
+                }
+                "absolute_consistency" => BlockstoreRecoveryMode::AbsoluteConsistency,
+                "point_in_time" => BlockstoreRecoveryMode::PointInTime,
+                "skip_any_corrupted_record" => BlockstoreRecoveryMode::SkipAnyCorruptedRecord,
+                _ => panic!("should not be allowed here"),
```
NIT: This would make a good `FromStr` impl on `BlockstoreRecoveryMode`

sakridge commented:
> 
> 
> LGTM! Just one nit
> 
> FYI, there's a DB to test this against on `ba3`, [#10015 (comment)](https://github.com/solana-labs/solana/issues/10015#issuecomment-651183798)

Cool, I added the option to solana-ledger-tool and also implemented `From<&str>` to share the logic. I confirmed the ba3 database can be opened with `--wal-recovery-mode skip_any_corrupted_record` without errors.

ryoqun commented:
@sakridge I think we can finally revisit this workaround fix for the `Corruption: SST file is ahead of WALs` errors from rocksdb.

Specifically, this relevant fix is finally landed in upstream and available in our codebase starting from: #18904

https://github.com/rust-rocksdb/rust-rocksdb/pull/531#issuecomment-874164361
